src/cephadm/cephadm

```diff
@@ -3450,26 +3453,159 @@ def command_list_networks():
+            return listen_address.split(':')[-1]
+        return None
+
+    def _get_grafana_port(file_name: str) -> Optional[str]:
+        with open(file_name) as f:
+            data = f.read().strip()
+        for line in data.split('\n'):
+            if 'http_port' in line and line.startswith(' '):
+                return line.split('=')[1].strip()
+        return None
+
+    def _get_web_port(service: str) -> Optional[str]:
+        dirs = os.listdir(os.path.join(data_dir, fsid))
+        for dir_name in dirs:
+            if dir_name.startswith(service):
+                if service in ['node-exporter', 'alertmanager', 'prometheus']:
+                    return _get_listen_address(os.path.join(data_dir, fsid, dir_name, 'unit.run'))
+                elif service == 'grafana':
+                    return _get_grafana_port(os.path.join(data_dir, fsid, dir_name, 'etc', 'grafana', 'grafana.ini'))
+                break
+        return None
+
+    # To get the version we're opting to use the daemon's API (if it has one), which is typically
+    # 10x faster than running a "exec <ID> <binary> --version" against the running container. We 
+    # only really use the "exec" form as a last resort to keep the runtime as low as possible
+    def _get_grafana_version(default_port: str = '3000') -> str:
+        port = _get_web_port('grafana') or default_port
+        ctx = ssl.create_default_context()
+        ctx.check_hostname = False
+        ctx.verify_mode = ssl.CERT_NONE
+        data = _http_call(
+            f"https://localhost:{port}/api/health",
+            ctx)
+        return json.loads(data)['version']
+
+    def _get_prometheus_version(default_port: str = '9090') -> str:
+        port = _get_web_port('prometheus') or default_port
+        data = _http_call(f"http://localhost:{port}/api/v1/status/buildinfo")
+        r = json.loads(data)
+        return r['data']['version']
+
+    def _get_node_exporter_version(default_port: str = '9100') -> str:
+        port = _get_web_port('node-exporter') or default_port
+        data = _http_call(f"http://localhost:{port}/metrics")
+        for metric in data.split('\n'):
+            if metric.startswith('node_exporter_build_info'):
+                # node_exporter_build_info{branch="HEAD",goversion="go1.12.5",revision="3db77732e925c08f675d7404a8c46466b2ece83e",version="0.18.1"} 1'
+                labels_str = re.split('[{}]', metric)[1]  # k=v,k=v,k=v
+                for kv_str in labels_str.split(','):
+                    k, v = kv_str.split('=')
+                    if k == 'version':
+                        return v.strip('"')
+        return ''
 
-    data_dir = args.data_dir
+    def _version_from_tag(image_name: str) -> str:
+        # if the tag looks like a version, default to that!
+        tag = image_name.split(':')[-1]
+        if re.search('^v?\d\.\d', tag):
+            return tag
+        else:
+            return ''
+    
+    def _get_tag(image_name: str) -> str:
```

Comment:
This is unused.

Also there is no safeguard against `image_name` being None.

Reply:
ACK

Reply:
removed